### DISCUSSION TITLE
mutation stuck in loading status after error if using onError

### DISCUSSION THREAD

**Discussion by ryanmk54** [TanStack/query#1484] [Q&A] [answered]
I have a mutation that gets stuck in loading status after the mutation function throws an error, but it only gets stuck if I provide an onError function. The onError function shows a notification in an ancestor component using the following code.
```
import {useCallback} from 'react';
import NestedError from '../helpers/NestedError';

const useNestedErrorNotification = ({notificationValues}) =>
  useCallback(
    (error: NestedError) => {
      notificationValues.handleOpen(error.toString(), 'error');
      throw (error.cause || error);
    },
    [notificationValues]
  );
```
I thought about creating an issue, but I haven't been able to make a minimal-complete-verifiable-example yet. One interesting thing to note is that when I tried making a code sandbox to reproduce the issue, it crashes the page, but when the error is thrown in my app. It doesn't crash, with and without the onError. I'm using rails-ranger (Axios) for api requests. I don't know if that would make a difference on the error behavior. This behavior happened in v2. I upgraded the mutations to react-query v3, loading status is still stuck.

Does anyone have any ideas why the loading status would be stuck when I include onError, or why the error doesn't crash the page? I thought I might be handling the error in a catch somewhere, but it doesn't crash without onError too. Is there something in Axios that prevents an error response from crashing the page?

**Comment by TkDodo** [comment, marked as answer]
am I reading this correctly that the above code is your `onError` handler for `useMutation` ? If so, I don't quite understand why you would _throw_ inside that error handler? That shouldn't be necessary - react-query knows about your error already, you can show the notification in `onError`, and you would still see the error + error status in the mutation...

**Reply by ryanmk54** [reply]
I was throwing so the error would print to the console. It seems like console.error wasn't printing the error object before. I commented out the throw. It still prints the error to the console and the mutation status doesn't get stuck on loading. Thanks! Is the react-query error status somehow tied to the onError function. I was using mutate in react-query v2 and mutateAsync in react-query v3

**Reply by TkDodo** [reply]
If you use mutateAsync i think you need try catch around it to not get an unhandled promise exception, but I’m not sure. In any case i wouldn’t throw if you are not catching it yourselves.

**Reply by toniton** [reply]
In a scenario, where you need to call another function in the `onError` callback that could potentially throw an error, and you may want to handle that with a generic error. 

How would you recommend handling that in `react-query`?

**Reply by ryanmk54** [reply]
Could you have a generic callback that handles the error? If you needed specific behavior you could pass a function into the callback. A code sample would be helpful to give more details.

**Reply by toniton** [reply]
@ryanmk54 Here is a snippet of the code for what I'm trying to achieve:

```js
const useExample = () => {
    const mutateResult = useMutation(
        async () => {

            const { data: response } = await axios.get(
                `https://example.com`,
            )

            if (response === "OK_BUT_NOT_ALRIGHT") {
                throw CustomError("help me feel alright!")
            }
            return { data: response }
        },
        {
            onError: (err) => {
                // Convert progressevent error to custom error. 
                if (err instanceof ProgressEvent) {
                    throw CustomError("interet seems to have been cut off!")
                }

                if (err instanceof CustomError) {
                    throw err
                }

                throw CustomError("Generic error for every other error!")
            },
        },
    )

    return mutateResult
}


const exampleMutate = useExample()
exampleMutate
    .mutateAsync()
    // I want to return only custom errors in catch here and use this 
    // always for showing the error modal.
    .catch((err) => showErrorModal(err).catch(console.log))
```

NB: `useMutation` is intentionally used, because I want to trigger the call from the click of a button.

**Reply by ryanmk54** [reply]
I would have an error handler function that wraps a function in a try catch. You could call it displayError. Something like:
```
const catchAndDisplayErrors = useCatchAndDisplayErrors()
...
onError: catchAndDisplayErrors((err) => {
    // Convert progressevent error to custom error. 
    if (err instanceof ProgressEvent) {
        throw CustomError("interet seems to have been cut off!")
    }

    if (err instanceof CustomError) {
        throw err
    }

    throw CustomError("Generic error for every other error!")
})
```
catchAndDisplayErrorrs could wrap the given function in a try catch and display the error with something like the useSnackbar hook from [notistack](https://github.com/iamhosseindhv/notistack).

**Reply by TkDodo** [reply]
> // Convert progressevent error to custom error.

why would you do that in `onError` rather than in the `queryFn` itself?

> NB: useMutation is intentionally used, because I want to trigger the call from the click of a button.

That's no reason to `useMutation` over `useQuery`. You'd likely want client state to drive the query via the key and set that client state when the button is clicked ;)

**Reply by ryanmk54** [reply]
> > // Convert progressevent error to custom error.
> 
> why would you do that in `onError` rather than in the `queryFn` itself?

Doing it in the queryFn would be better, as it would be handled sooner there. Also if using a [default queryFn](https://react-query.tanstack.com/guides/default-query-function), there is no need to add a custom error handler to every onError.

Converting to a custom error would be better done in the queryFn, but I find it easier to display the error in onError, because I have access to the notification hook there.

**Reply by TkDodo** [reply]
onError is not meant for transforming errors, same as onSuccess can't transform data. They are functions to execute side-effects and don't return anything.

**Reply by toniton** [reply]
> onError is not meant for transforming errors, same as onSuccess can't transform data. They are functions to execute side-effects and don't return anything.

@TkDodo. 
It looks like network requests was intercepted by another function, and network errors were not being triggered at the query function level. Instead, network errors were returned as `ProgressEvent` in the onError callback. Hence why I needed the custom error handler. 

_After setting up a bare react-query installation, I get a better error._

**Reply by toniton** [reply]
> That's no reason to `useMutation` over `useQuery`. You'd likely want client state to drive the query via the key and set that client state when the button is clicked ;)

My actual useCase:
`onClick of button ---> call API A (check status) ---> onPrecondition passed ---> call API B (perform action).`

I find it more convenient using `useMutation` over `useQuery` to make on-demand API `GET` calls that does not require all the perks of a query, as I won't have to manage states to make queries. Also, in this case, query keys and other sorts of caching are not needed for checking status since I want to make a new call everytime the button is clicked.

A button to check the status of an item would be one of such examples where `mutateAsync.then` takes preference over `enabled: variable` for me. 😉

**Comment by igortas** [comment]
@TkDodo 
I still having the problem, returning errors only from the fetcher, executes both ```onError``` and ```onSuccess```.
It's weird that RQ don't handle well first fail scenario, ```isLoading``` is stuck to ```true```.

**Reply by ryanmk54** [reply]
A code snippet would be helpful here.
